drivers/storage/ide/atapi/satl.c

```diff
@@ -0,0 +1,1459 @@
+        TaskFile->HighLba = (UCHAR)(Sector >> 8);
+        TaskFile->DriveSelect = (UCHAR)Head;
+    }
+}
+
+static
+ULONG
+AtaReqReadWrite(
+    _In_ PATAPORT_CHANNEL_EXTENSION ChannelExtension,
+    _In_ PATAPORT_DEVICE_EXTENSION DeviceExtension,
+    _In_ PSCSI_REQUEST_BLOCK Srb)
+{
+    PATA_DEVICE_REQUEST Request = &ChannelExtension->Request;
+    PCDB Cdb;
+    ULONG64 Lba;
+    ULONG SectorCount;
+
+    Request->Complete = AtaReqCompleteNoOp;
+    Request->DeviceExtension = DeviceExtension;
+    Request->DataBuffer = Srb->DataBuffer;
+    Request->DataTransferLength = Srb->DataTransferLength;
+    Request->Mdl = Request->Irp->MdlAddress;
+    Request->Flags = REQUEST_FLAG_ASYNC_MODE |
+                     REQUEST_FLAG_SET_DEVICE_REGISTER |
+                     REQUEST_FLAG_READ_WRITE;
+
+    Cdb = (PCDB)Srb->Cdb;
+    switch (Cdb->AsByte[0])
+    {
+        case SCSIOP_READ6:
+        case SCSIOP_WRITE6:
+        {
+            Lba = CdbGetLogicalBlockAddress6(Cdb);
+            break;
+        }
+
+        case SCSIOP_READ:
+        case SCSIOP_WRITE:
+        {
+            Lba = CdbGetLogicalBlockAddress10(Cdb);
+
+            if (Cdb->CDB10.ForceUnitAccess)
+                Request->Flags |= REQUEST_FLAG_FUA;
+            break;
+        }
+
+        case SCSIOP_READ12:
+        case SCSIOP_WRITE12:
+        {
+            Lba = CdbGetLogicalBlockAddress12(Cdb);
+
+            if (Cdb->CDB12.ForceUnitAccess)
+                Request->Flags |= REQUEST_FLAG_FUA;
+            break;
+        }
+
+        case SCSIOP_READ16:
+        case SCSIOP_WRITE16:
+        {
+            Lba = CdbGetLogicalBlockAddress16(Cdb);
+
+            if (Cdb->CDB16.ForceUnitAccess)
+                Request->Flags |= REQUEST_FLAG_FUA;
+            break;
+        }
+
+        default:
+            ASSERT(FALSE);
+            UNREACHABLE;
+    }
+
+    /* Check for write operations */
+    if (Cdb->AsByte[0] & 0x02)
+        Request->Flags |= REQUEST_FLAG_DATA_OUT;
+    else
+        Request->Flags |= REQUEST_FLAG_DATA_IN;
+
+    if (!(DeviceExtension->Flags & DEVICE_PIO_ONLY))
+    {
+        Request->Flags |= REQUEST_FLAG_DMA_ENABLED;
+    }
+    if (DeviceExtension->MultiSectorTransfer != 0)
+    {
+        Request->Flags |= REQUEST_FLAG_READ_WRITE_MULTIPLE;
+    }
+
+    /* FIXME */
+    Request->Flags &= ~REQUEST_FLAG_FUA;
+
+    /* Ignore the transfer length field in the CDB, use the DataTransferLength field instead */
+    SectorCount = Request->DataTransferLength + (DeviceExtension->SectorSize - 1);
+    SectorCount /= DeviceExtension->SectorSize;
+
+    /* Make sure the sector count do not exceed the maximum transfer length */
+    ASSERT(SectorCount <= 0x100);
+
+    AtaReqBuildLbaTaskFile(ChannelExtension, DeviceExtension, Lba, SectorCount);
+
+    /* Choose the command opcode */
+    Request->TaskFile.Command = AtaReadWriteCommand(Request, DeviceExtension);
+
+    /* Terminate the request when the FUA is not supported */
```

Comment:
What is a FUA?

Reply:
https://en.wikipedia.org/wiki/Disk_buffer#FUA